cc/interactions: document interactions (#31)
botlabs-gg/yagpdb-docs-v2: content/reference/custom_interactions.md

+++
title = 'Custom Interactions'
weight = 6
+++

Buttons, Modals, Select Menus, Ephemeral Messages, and more!

<!--more-->

{{% notice warning %}}

Use of interactions within YAGPDB is an advanced topic. You will need a thorough understanding of YAGPDB's scripting
language before learning interactions.

> We have a comprehensive learning course located [here](https://learn.yagpdb.xyz) for you to get started.

Reviewer: Wanted to note this down as a general style suggestion: we should avoid using "click here" or "located here" in links; see, e.g., <https://ux.stackexchange.com/questions/12100/why-shouldnt-we-use-words-such-as-here-and-this-in-textlinks> for reasoning. In this case I would prefer to write
```suggestion
We have a [comprehensive learning course](https://learn.yagpdb.xyz) for you to get started.
```

Reviewer 2: Thanks, in my initial review I somehow missed that. I think we have a lot of similar things in the migrated documentation files as well; I will review those at a later date.

{{% /notice %}}

## The Basics

> Interactions within Discord allow server members to use alternative, built-in features to trigger bots to take action
> other than messages or reactions. These features include builtin buttons, dropdown selection menus, or submitting a
> modal (basically a pop-up form). Within custom commands it is possible to both create and customize these new
> interactive features, but respond to them as well, opening up new possibilities for ephemeral message responses, modals,

Reviewer: ```suggestion
Interactions within Discord allow server members to use alternative, built-in features to trigger bots to take action
other than messages or reactions. These features include builtin buttons, dropdown selection menus, or submitting a
modal (basically a pop-up form). Within custom commands it is possible to not only create and customize these new
interactive features, but respond to them as well, opening up new possibilities for ephemeral message responses, modals,
```

and more within YAGPDB custom templating script.

### Definitions

**Interaction** - A user engaging with YAGPDB through one of Discord's builtin features: Clicking a button, Making a
selection with a select menu, or Submitting a modal.

> **Response** - YAGPDB is required to send an initial interaction response after receiving an *interaction*. If it does
> not do this, the user triggering the interaction will see a "This application did not respond" error. The bot cannot
> respond to an interaction more than once. YAGPDB may only respond to an interaction in one of the following ways:
> Sending a message, sending a modal, or updating the message which the *interaction* was triggered on.

Reviewer: ```suggestion
**Response** - YAGPDB is required to respond promptly after receiving an interaction by either sending a message or modal, or by updating the message on which the interaction was triggered. If it does not do this, the user triggering the interaction will see a "This application did not respond" error. The bot cannot respond to an interaction more than once.
```
This reformulation puts the definition of "initial interaction response" up front, which I think is preferable.

**Followup** - Since YAGPDB may only *respond* to an *interaction* once, it is subsequently required to send an interaction
followup if it still needs to respond to the interaction. These followups can be sent up to 15 minutes after the initial
interaction, and you can send as many as you want. YAGPDB may only send a followup in one of the following ways: Sending
a message, editing a response or followup, or getting a response or followup message.

> 

Reviewer: This is confusingly worded. "YAGPDB may only send a followup in one of the following ways: [...] editing a followup, or getting [...] a followup" (?)

Author: Maybe 'editing/getting a previous followup message'?

**Interactive Elements** - Elements users can interact with to send *interactions*, i.e. buttons, menus, and modals.

> **Message Components** - *Interactive Elements* which can be attached to YAGPDB's Discord messages, i.e. buttons and menus.

Reviewer: I understand these are the terms that the Discord API uses, but I would rather inline these terms into their use sites. So instead of saying "Creating message components" I would much rather see "Creating buttons and menus"; this seems more amenable to skimming (and, in general, more appropriate for an audience that does not work with the Discord API directly.)

Author: I'm trying to get people familiar with the idea of message components for a few reasons:

1. A component of a message makes it clear that these are just parts of any regular message, not just interaction responses. During prototyping a few users struggled to understand how to create buttons or menus in the first place.
2. The CC trigger type is a "message component" type.

Reviewer 2: I think this is fine if we write the other reference pages in a similar style, where appropriate.
Otherwise I agree with Joe -- in-depth explanations belong, in my opinion, to learn.yagpdb.

For now I propose we keep this as-is and then once we migrate the learning resources over to Hugo as well we move details over there. What do you think?

buttons' appearance and behavior with color, emoji, label text, etc. When a button is clicked, an *interaction* is sent
to the bot.

**Menu** - A dropdown select menu appearing in or under a Discord message sent by YAGPDB. You can create and customize these
menus' appearance and behavior with placeholder text, predefined options with labels, descriptions, and/or emojis,
designate the entire menu as a user or role select menu instead, etc. When a select menu is used, an *interaction* is sent
to the bot.

**Modal** - A pop-up form YAGPDB can send in response to an interaction. It allows users to privately input text which
is sent directly to YAGPDB for use in CC scripting. You can create and customize these modals' appearance and
behavior with a title and fields. YAGPDB can both **receive a submitted modal** (which is an
*interaction*), and **send a modal** for a member to fill out, (which is an interaction *response*).

**Ephemeral** - An ephemeral message is sent to a server channel but only appears to a single user. YAGPDB cannot send
these ephemeral messages to users except in response to an *interaction*. Both *response* messages and *followup*
messages can be ephemeral.

> ## Interaction Lifetime

Reviewer: This is a nice summary (with a visual as well!) I would suggest it be placed before the definition as a sort of "basic intro". In general I believe guides or tutorials (which this article appears to be, as opposed to reference documentation per se) should begin with a brief overview that still has some usable technical detail before diving into specific definitions and so on.

{{ $message := complexMessage "buttons" $button }}
{{ sendMessage nil $message }}
```

This button triggers our duck RegEx custom command.

Now, let's add some more buttons.

{{% notice info %}}

You cannot have two message components with the same Custom ID in one message.

{{% /notice %}}

{{% notice info %}}

Buttons with the "link" style cannot have a Custom ID, and instead require a URL field.

Link style buttons do not trigger *interactions*.

{{% /notice %}}

```go
{{ $button1 := cbutton "label" "Duck One" "custom_id" "buttons-duck-alpha" "style" "success" }}
{{ $button2 := cbutton "emoji" (sdict "name" "🦆") "custom_id" "buttons-duck-beta" "style" "danger" }}
{{ $button3 := cbutton "label" "Duck Three" "emoji" (sdict "name" "🦆") "url" "https://yagpdb.xyz" "style" "link" }}
{{ $message := complexMessage "buttons" (cslice $button1 $button2 $button3) }}
{{ sendMessage nil $message }}
```

![Many buttons](many_buttons.png)

At this stage we have three buttons. Both of the first two buttons will trigger our duck trigger custom command, but the
third button will not trigger any custom command. Link buttons do not create *interactions*.

> We can differentiate between the two buttons with `.StrippedID`, which returns the `.CustomID` but with the trigger and
> everything before it (in our case, `buttons-duck`) stripped off.

Reviewer 2: ```suggestion
We can differentiate between the two buttons using `.StrippedID`, which, just like `.StrippedMsg`, returns our Custom ID
without the trigger and everything else before that. In our example, `.StrippedID` will return `-alpha` for the first
button and `-beta` for the second button.

Confirming this behavior will be left as an exercise to the reader (you).
```

What do you think about this rephrase?

{{ $menu := cmenu
  "type" "text"
  "placeholder" "Choose a terrible thing"
  "custom_id" "menus-duck-alpha"
  "options" (cslice
    (sdict "label" "Ducks" "value" "opt-1" "default" true)
    (sdict "label" "Duck" "value" "opt-2" "emoji" (sdict "name" "🦆"))
    (sdict "label" "Half a Duck" "value" "opt-3" "description" "Don't let the smaller amount fool you."))
  "max_values" 3 }}

{{ $message := complexMessage "buttons" (cslice $button1 $button2 $button3) "menus" $menu }}
{{ sendMessage nil $message }}
```

![Buttons and menus](buttons_and_menus.png)

We now have two buttons and a menu which are triggering our duck custom command. We used to branch with `.StrippedID`,
but now since we have two components whose stripped IDs are `-alpha`, we'll need to branch with `.IsMenu` too. If our
custom command was triggered by the use of this select menu, we could also see that `.MenuType` returns `"string"`,
since this menu is a string select type menu.

##### Ordering Message Components

A message can have 5 rows of components.
A row of components can have 5 buttons, **or** 1 menu.

> Let's say I wanted to play tic tac toe. If I just add 9 buttons into the same slice in my complex message, they'll just

Reviewer 2: ```suggestion
Let's say we want to play tic tac toe. If we just add 9 buttons into the same slice in my complex message, they'll just
```

I'm assuming this a cookbook style you're aiming for, "we" seems more appropriate.

{{ if $validCard }}
  {{ sendResponse nil (print .User " played a " $playedCard.Color $playedCard.Number) }}
  {{ dbSet .Channel.ID "uno-last-card" $playedCard }}
{{ else }}
  {{ sendResponse nil "You can't play that card!" }}
{{ end }}
```

Example 2: A user is setting up a new UNO game with a modal, they've filled out a 'number of decks in play' and a
'minimum number of cards to play' field, triggering the following command which will update those values in database.

![A modal for setting up a game of UNO](uno-modal.png)

```go
{{ $numberOfDecks := index .Values 0 }}
{{ $minCardsForUNO := index .Values 1 }}

{{ dbSet .Channel.ID "uno-decks" ( toInt $numberOfDecks ) }}
{{ dbSet .Channel.ID "uno-min_for_uno" ( toInt $minCardsForUNO ) }}
```

## Responding to an Interaction

### Initial Response

While technically not required, responding to an interaction with one of Discord's allotted initial responses is crucial
if you don't want your users to see an error after interacting. An interaction may be responded to only once.

You can only respond to an interaction within the custom command triggered by said interaction, with the exception that

> a CC executed with `execCC` by the triggered CC will be able to send initial responses to the triggering interaction.

Reviewer 2: ```suggestion
a CC executed with `execCC` by the triggered CC will be able to send initial responses to the triggering interaction as well.
```

```go
{{ $rows := cslice }}
{{ $enemies := cslice "Dragon" "Ogre" "Duck" }}

{{ range $enemyName := $enemies }}
  {{ $nameButton := cbutton "label" $enemyName "style" "grey" "disabled" true }}
  {{ $attackButton := cbutton "label" "Attack" "style" "red" "custom_id" (print "attack-" (lower $enemyName)) }}
  {{ $befriendButton := cbutton "label" "Befriend" "style" "green" "custom_id" (print "befriend-" (lower $enemyName)) }}

  {{ $currentRow := cslice $nameButton $attackButton $befriendButton }}
  {{ $rows = $rows.Append $currentRow }}
{{ end }}

{{ $promptText := joinStr ", " "Suddenly" $enemies "attack you!" }}
{{ $message := complexMessage "content" $promptText "components" $rows }}
{{ sendMessage nil $message }}
```

![Result of the Full Solution Code](full-variable-row-solution.png)

#### Emojis in Message Components

Buttons and Select Menu Options both have an `"emoji"` field, but this field does not accept the regular unicode/name:id
formula like reactions do. Emojis in components follow the [partial emoji
object](https://discord.com/developers/docs/resources/emoji#emoji-object) structure, however only the ID *or* the Name

> fields are required, depending on if you are using a custom emoji or not.

Author: Yes I will update my code examples to reflect that only an ID is
required when using Custom IDs. I have verified through testing
that this is the case, only an ID is needed for custom emojis and
regardless of if it is animated or not the animated field is not
necessary.

{{ if $validCard }}
  {{ sendResponse nil (print .User " played a " $playedCard.Color $playedCard.Number) }}
  {{ dbSet .Channel.ID "uno-last-card" $playedCard }}
{{ else }}
  {{ sendResponse nil "You can't play that card!" }}
{{ end }}
```

Example 2: A user is setting up a new UNO game with a modal, they've filled out a 'number of decks in play' and a
'minimum number of cards to play' field, triggering the following command which will update those values in database.

![A modal for setting up a game of UNO](uno-modal.png)

```go
{{ $numberOfDecks := index .Values 0 }}
{{ $minCardsForUNO := index .Values 1 }}

{{ dbSet .Channel.ID "uno-decks" ( toInt $numberOfDecks ) }}
{{ dbSet .Channel.ID "uno-min_for_uno" ( toInt $minCardsForUNO ) }}
```

## Responding to an Interaction

### Initial Response

While technically not required, responding to an interaction with one of Discord's allotted initial responses is crucial
if you don't want your users to see an error after interacting. An interaction may be responded to one time.

You can only respond to an interaction within the custom command triggered by said interaction. A CC executed with

> `execCC` by the triggered CC will be able to send initial responses to the triggering interaction.

Author: Does this change adequately address?

  - You can also use this to send `embeds` or `complexMessages`.
  - You'll need to send a `complexMessage` and pass it `"ephemeral" true` as an argument to send *ephemeral* messages.
  - `sendResponse` comes in `NoEscape` and `RetID` variants too.
  - When sending an initial response, `sendResponse` does not need an interaction token, `nil` can be used.
- Use the `sendModal` function to show the user a modal. You cannot respond to a user submitting a modal by sending them
  another modal.
- Use the `updateMessage` function to edit the message the command triggered from. This works the same way as editing a
  message, however because it automatically targets the triggering message, the only argument required is the new
  message.

[Interaction Function documentation](/reference/templates/functions#interactions)

### Following Up

Followups allow you to continue responding to an interaction after the initial response has been made. You can followup
for up to 15 minutes after the user interacts, and you can follow up as many times as you'd like. Followups require the
interaction token of the interaction they should be following up on.

> Possible initial responses:

Author: Sorry that was just me forgetting to edit the title after copy pasting from
the previous section, thank you for catching that.

parsing options from mentionable type menus, you will need to use your own methods of determining if an ID is a role or
a user.

**Default Values**

Setting default values in these select menus is a more involved process than for text type menus. Instead of setting a
`default` value on each option, you must instead provide a `default_values` argument containing a slice of [default
value
structures](https://discord.com/developers/docs/interactions/message-components#select-menu-object-select-default-value-structure).

```go
{{ $adminRoleID := "1210128415689285642" }}
{{ $soggysaussagesUserID := "329826270898683904" }}

{{ $menu := cmenu
  "type" "mentionable"
  "placeholder" "Choose users or roles who are secretly ducks"
  "custom_id" "menus-duck-mentionables"
  "default_values" ( cslice
  ( sdict "type" "role" "id" $adminRoleID )
  ( sdict "type" "user" "id" $soggysaussagesUserID ))
  "max_values" 3 }}

{{ sendMessage nil (complexMessage "menus" $menu) }}
```

![The above mentionable menu with default options selected](mentionable_menu_with_defaults.png)

{{% notice info %}}

> The `default_values` `id`s must be `strings`, not `int64s`. Note how in the above example, `$adminRoleID :=

Author: I totally agree with this and actually did this initially but thought that might not read well and did it this way instead :ye: Thanks for affirmation that that is the correct way to go
